projects/storefrontlib/src/cms-components/checkout/guards/checkout-steps-set.guard.ts

```diff
@@ -0,0 +1,173 @@
+import { Injectable, isDevMode } from '@angular/core';
+import {
+  CanActivate,
+  UrlTree,
+  ActivatedRouteSnapshot,
+  RouterStateSnapshot,
+  Router,
+} from '@angular/router';
+import { Observable, of, combineLatest } from 'rxjs';
+import { map, tap, switchMap, take } from 'rxjs/operators';
+import {
+  Address,
+  RoutingConfigService,
+  PaymentTypeService,
+  CheckoutCostCenterService,
+} from '@spartacus/core';
+import { CheckoutStepService } from '../services/checkout-step.service';
+import { CheckoutDetailsService } from '../services/checkout-details.service';
+import { CheckoutStep, CheckoutStepType } from '../model/checkout-step.model';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class CheckoutStepsSetGuard implements CanActivate {
+  constructor(
+    private paymentTypeService: PaymentTypeService,
+    private checkoutStepService: CheckoutStepService,
+    private checkoutDetailsService: CheckoutDetailsService,
+    private routingConfigService: RoutingConfigService,
+    private checkoutCostCenterService: CheckoutCostCenterService,
+    private router: Router
+  ) {}
+
+  canActivate(
+    route: ActivatedRouteSnapshot,
+    _: RouterStateSnapshot
+  ): Observable<boolean | UrlTree> {
+    let currentIndex = -1;
+    const currentRouteUrl = '/' + route.url.join('/');
+
+    // check whether the previous step is set
+    return combineLatest([
+      this.checkoutStepService.steps$,
+      this.paymentTypeService.isAccountPayment(),
+    ]).pipe(
+      tap(([, isAccount]) => {
+        this.checkoutStepService.disableEnableStep(
+          CheckoutStepType.PAYMENT_DETAILS,
+          isAccount
+        );
+      }),
+      take(1),
+      switchMap(([steps, isAccount]) => {
+        steps.forEach((step, index) => {
+          const stepRouteUrl = `/${
+            this.routingConfigService.getRouteConfig(step.routeName).paths[0]
+          }`;
+          if (stepRouteUrl === currentRouteUrl) {
+            currentIndex = index;
+          }
+        });
+        if (this.validateCurrentStep(currentIndex, steps, currentRouteUrl)) {
+          return this.isPreviousStepSet(steps[currentIndex - 1], isAccount);
+        } else {
+          return of(this.getUrl('checkout'));
+        }
+      })
+    );
+  }
+
+  protected validateCurrentStep(
+    currentIndex: number,
+    steps: CheckoutStep[],
+    routeUrl: string
+  ): boolean {
+    let currentStep: CheckoutStep;
+    if (currentIndex >= 0) {
+      currentStep = steps[currentIndex];
+    }
+    if (!currentStep && isDevMode()) {
+      console.warn(
+        `Missing step with route '${routeUrl}' in checkout configuration.`
+      );
+      return false;
+    }
+    return true;
+  }
+
+  protected isPreviousStepSet(
```

Comment:
Suggestion.  `isStepSet` might improve readability, since the function just takes whatever step is passed as an argument and checks if it's set.  Your call.